### PR TITLE
Cpm-634 Create Get Product by uuid endpoint 

### PR DESCRIPTION
**Pull request by amigaire**
**Description (for Contributor and Core Developer)**

Create a new endpoint to get a product by its uuid: a new Controller has been created for the GetProduct action

**Definition Of Done (for Core Developer only)**

- [x] Tests

### CODE REVIEW

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/ValueObject/ProductUuid.php [hunk 1]
```diff
@@ -8,7 +8,7 @@
 use Ramsey\Uuid\UuidInterface;
 
 /**
- * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Domain/ValueObject/ProductUuid.php:11]: rollback please

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Apps/Persistence/FindOneConnectedAppByUserIdQueryIntegration.php [hunk 2]
```diff
@@ -40,7 +40,7 @@ public function test_it_returns_connected_app_associated_to_the_user(): void
             'outdated_app_id',
             'outdated_app_code',
         );
-        $userId = $this->getUserId('outdated_app_code');
+        $userId = $this->getUserIdFromFirstname('outdated_app_code');
```
Comment (nmarniesse) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Apps/Persistence/FindOneConnectedAppByUserIdQueryIntegration.php:43]: Why this change?

File: phpunit.xml.dist [hunk 3]
```diff
@@ -73,10 +73,6 @@
             <directory suffix="EndToEnd.php">src</directory>
         </testsuite>
 
-        <testsuite name="OroSecurityBundle">
-            <file>src/Oro/Bundle/SecurityBundle/Tests/Unit/Acl/Dbal/MutableAclProviderTest.php</file>
-        </testsuite>
-
```
Comment (nmarniesse) [inline phpunit.xml.dist:79]: Why?

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/Uuid/GetProductScoresByUuidsQuery.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        $query = <<<SQL
+SELECT BIN_TO_UUID(product.uuid) as uuid, latest_score.scores, latest_score.scores_partial_criteria
+FROM pim_catalog_product product
+INNER JOIN pim_data_quality_insights_product_score AS latest_score ON latest_score.product_uuid = product.uuid
+LEFT JOIN pim_data_quality_insights_product_score AS younger_score
+    ON younger_score.product_uuid = latest_score.product_uuid
+    AND younger_score.evaluated_at > latest_score.evaluated_at
+WHERE product.uuid IN(:product_uuids)
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/Infrastructure/Persistence/Query/ProductEvaluation/Uuid/GetProductScoresByUuidsQuery.php:48]:
```suggestion
WHERE product.uuid IN (:product_uuids) 
```

File: src/Akeneo/Pim/Automation/DataQualityInsights/tests/back/Specification/Domain/ValueObject/ProductUuidSpec.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace Specification\Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject;
+
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\ProductUuid;
+use PhpSpec\ObjectBehavior;
+use Ramsey\Uuid\Uuid;
+
+final class ProductUuidSpec extends ObjectBehavior
+{
+    // TODO tests kos
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Automation/DataQualityInsights/tests/back/Specification/Domain/ValueObject/ProductUuidSpec.php:18]: todo

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        private SecurityFacade $security,
+    ) {
+    }
+
+    /**
+     * @param Request $request
+     * @param string $uuid
+     * @return JsonResponse
+     */
+    public function getAction(Request $request, string $uuid): JsonResponse
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php:49]:
```suggestion
    public function __invoke(Request $request, string $uuid): JsonResponse
```

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        switch ($acl) {
+            case 'pim_api_product_list':
+                return 'Access forbidden. You are not allowed to list products.';
+            case 'pim_api_product_edit':
+                return 'Access forbidden. You are not allowed to create or update products.';
+            case 'pim_api_product_remove':
+                return 'Access forbidden. You are not allowed to delete products.';
+            default:
+                return 'Access forbidden.';
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php:98]: Can be simplify. Only `pim_api_product_list` is needed

**Review by nmarniesse — COMMENTED** [review]
Don't finish the review, but here is some comments.  
An idea, but I don't know if it's better, can we modify the existing query to not return by identifier but by uuid? Then in the actual API we just have to do a conversion identifiers -> uuid. (But mqybe the queries are used elsewhere so maybe not so easy)

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Apps/Persistence/FindOneConnectedAppByUserIdQueryIntegration.php (hunk 2 above)
Comment (amigaire) [reply src/Akeneo/Connectivity/Connection/back/tests/Integration/Apps/Persistence/FindOneConnectedAppByUserIdQueryIntegration.php:43]: I added a getUserId method in TestCase class which is extended by this integration test class.
The getUserId in TestCase get a user Id from the **username**, but in the current test file the method getUserIdFromFirstname was named getUserId and got the user id from the user **firstname**.
To avoid confusion, I renamed the getUserId method in this test class to getUserIdFromFirstname

I will change the tests to retrieve the userId by the username

Comment (amigaire) [reply src/Akeneo/Connectivity/Connection/back/tests/Integration/Apps/Persistence/FindOneConnectedAppByUserIdQueryIntegration.php:43]: During the test a user is created via $this->connectedAppLoader->createConnectedAppWithUserAndTokens(
            'outdated_app_id',
            'outdated_app_code',
        );
The username is generated with username= 'outdated_app_code<random 4 digit>' 

The TestCase::getUserId method use a query with the clause "where username = $userId" => so it cannot retrieve the previously created user.
This user first_name field has the same value as the code 'outdated_app_code', that's why we use a specific method in this class to retrieve the userId

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/Uuid/GetProductScoresQuery.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        private GetProductScoresByUuidsQuery $getProductScoresByUuidsQuery,
+        private GetScoresByCriteriaStrategy $getScoresByCriteria,
+    ) {
+    }
+
+    /**
+     * @param UuidInterface[] $productUuids
+     * @return QualityScoreCollection[]
+     */
+    public function byProductUuids(array $productUuids): array
```
Comment (pierallard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/Uuid/GetProductScoresQuery.php:32]: For me, we have to update https://github.com/akeneo/pim-community-dev/blob/6996b3d5f22aeb04afb9a9e58adc3164dbdfb589/src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/GetProductScoresQuery.php instead of adding a new class.
If we let this this way, DQI will not work for products without identifiers.

File: src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/Uuid/GetProductScoresQuery.php [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        $scores = $this->getProductScoresByUuidsQuery->byProductUuid($productUuid);
+
+        return $this->qualityScoreCollection(($this->getScoresByCriteria)($scores));
+    }
+
+    /**
+     * @param ChannelLocaleRateCollection $channelLocaleRateCollection
+     * @return QualityScoreCollection
+     */
+    private function qualityScoreCollection(ChannelLocaleRateCollection $channelLocaleRateCollection): QualityScoreCollection
```
Comment (pierallard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/back/PublicApi/Query/ProductEvaluation/Uuid/GetProductScoresQuery.php:57]: Useless method, move it back to the other one

File: src/Akeneo/Pim/Automation/DataQualityInsights/tests/back/Specification/Domain/ValueObject/ProductUuidSpec.php [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Specification\Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject;
+
+use Akeneo\Pim\Automation\DataQualityInsights\Domain\ValueObject\ProductUuid;
+use PhpSpec\ObjectBehavior;
+use Ramsey\Uuid\Uuid;
+
+final class ProductUuidSpec extends ObjectBehavior
+{
+    public function it_can_be_construct_from_a_string()
```
Comment (pierallard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/tests/back/Specification/Domain/ValueObject/ProductUuidSpec.php:18]:
```suggestion
    public function it_can_be_constructed_from_a_string()
```

File: src/Akeneo/Pim/Automation/DataQualityInsights/tests/back/Specification/Domain/ValueObject/ProductUuidSpec.php [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        $this->__toString()->shouldReturn('df470d52-7723-4890-85a0-e79be625e2ed');
+    }
+
+    public function it_throws_an_exception_if_the_product_id_is_not_string_when_using_fromString()
+    {
+        $id = 12;
+
+        $this->beConstructedThrough('fromString', [$id]);
+        $this->shouldThrow(\InvalidArgumentException::class)->duringInstantiation();
+    }
```
Comment (pierallard) [inline src/Akeneo/Pim/Automation/DataQualityInsights/tests/back/Specification/Domain/ValueObject/ProductUuidSpec.php:42]:
```suggestion
    public function it_throws_an_exception_if_the_product_uuid_is_not_string_when_using_fromString()
    {
        $this->beConstructedThrough('fromString', [12]);
        $this->shouldThrow(\InvalidArgumentException::class)->duringInstantiation();
    }
```

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+     * @param string $uuid
+     * @return JsonResponse
+     */
+    public function __invoke(Request $request, string $uuid): JsonResponse
+    {
+        $this->denyAccessUnlessAclIsGranted('pim_api_product_list');
+
+        $productUuid = Uuid::fromString($uuid);
+
+        $connectorProductsQuery = 'true' === $request->query->get('with_attribute_options', "false") ?
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php:55]:
```suggestion
        $connectorProductsQuery = 'true' === $request->query->get('with_attribute_options', 'false') ?
```

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/Uuid/SqlGetConnectorProducts.php [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+            $rows[$uuid->toString()]['raw_values'] = $rawValues;
+            $rawValuesIndexedByProductUuid[$uuid->toString()] = $rawValues;
+        }
+
+        $filteredRawValuesIndexedByProductUuid = $this->readValueCollectionFactory->createMultipleFromStorageFormat($rawValuesIndexedByProductUuid);
+
+        $products = [];
+        foreach ($productUuids as $uuid) {
+            if (!isset($rows[$uuid->toString()]['identifier'])) {
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/Uuid/SqlGetConnectorProducts.php:102]: Remove this condition

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/Uuid/SqlGetConnectorProducts.php [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+                $filteredRawValuesIndexedByProductUuid[$uuid->toString()],
+                null,
+                null
+            );
+        }
+
+        return new ConnectorProductList(count($products), $products);
+    }
+
+    private function removeIdentifierValue(array $rawValues, string $identifierAttributeCode): array
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Connector/Uuid/SqlGetConnectorProducts.php:129]: True question ; maybe we should not remove identifier value here ? To prepare the subject of "multiple identifiers" in the next months (it has no sense to remove it for me)
